docs/dev-process/coding-guidelines.md

```diff
@@ -1,92 +1,172 @@
-because code and patterns are copied we want certain inefficient code to stay out of the performance critical code.
+We have some guidelines that we typically apply widely even in less important code because code and patterns are copied,
+and we want certain inefficient code to stay out of the performance critical code.
 
 Some general guidelines:
 
-* Avoid LINQ - it can create lots of avoidable garbage
+* Avoid LINQ - it can create lots of avoidable garbage.
+
 * Prefer `for` and `foreach`,
   with a slight preference towards `for` when you're uncertain if `foreach` allocates an iterator.
+
 * Avoid `params` arrays, prefer adding overloads with 1, 2, 3, and maybe more parameters.
+
 * Be aware of APIs such as `String.Split(params char[])` that do not provide overloads to avoid array allocation.
-  When calling such APIs, reuse a static array when possible.
+  When calling such APIs, reuse a static array when possible (i.e. `Utils.Separators.Colon`).
+
+* Avoid creating empty arrays.
+  Instead, reuse the static ones via `Utils.EmptyArray<T>`.
+
 * Avoid unnecessary memory allocation in a loop.
   Move the memory allocation outside the loop if possible.
 
-## Portable code
+* Use `dict.TryGetValue` instead of `dict.Contains` and `dict[<key>]` when retrieving value from a `Dictionary`.
+  In this way you can avoid hashing the key twice.
+
+## Security Considerations
+
+Security is an important aspect of PowerShell and we need to be very careful about changes that may introduce security risks.
+Reviewers of a PR should be sensitive to changes that may affect security.
+Some security related keywords may serve as good indicators,
+such as `crypto`, `encryption`, `decryption`, `certificate`, `authenticate`, `ssl/tls` and `protected data`.
+
+When facing a PR with such changes,
+the reviewers should request a designated security Subject Matter Expert (SME) to review the PR.
+Currently, @PaulHigin and @TravisEz13 are our security SMEs.
+See [CODEOWNERS](../../.github/CODEOWNERS) for more information about the area experts.
+
+## Best Practices
+
+* Avoid hard-coding anything unless it's absolutely necessary.
 
-The PowerShell code base started on Windows and depends on many Win32 APIs through P/Invoke.
-Going forward, we try to depend on CoreCLR to handle platform differences,
-so avoid adding new P/Invoke calls where a suitable alternative exists in .NET.
+* Avoid a method that is too long and complex.
+  In such case, separate it to multiple methods or even a nested class as you see fit.
+
+* Use `using` statement instead of `try/finally` if the only code in the `finally` block is to call the `Dispose` method.
+
+* Use of object initializers (i.e. `new Example { Name = "Name", ID = 1 }`) is encouraged for better readability,
+  but not required.
+
+* Stick to the `DRY` principle -- Don't Repeat Yourself.
+   * Wrap the commonly used code in methods, or even put it in a utility class if that makes sense,
```

Comment:
Maybe mention some of the existing utility classes.

Reply:
Add an example utility class `System.Management.Automation.StringToBase64Converter`